src/material/badge/badge.ts

```diff
@@ -145,70 +145,54 @@ export class MatBadge extends _MatBadgeBase implements OnDestroy, OnChanges, Can
     return this.position.indexOf('before') === -1;
   }
 
-  ngOnChanges(changes: SimpleChanges) {
-    const contentChange = changes['content'];
-
-    if (contentChange) {
-      const value = contentChange.currentValue;
-      this._hasContent = value != null && `${value}`.trim().length > 0;
-      this._updateTextContent();
-    }
-  }
-
-  ngOnDestroy() {
-    const badgeElement = this._badgeElement;
-
-    if (badgeElement) {
-      if (this.description) {
-        this._ariaDescriber.removeDescription(badgeElement, this.description);
-      }
-
-      // When creating a badge through the Renderer, Angular will keep it in an index.
-      // We have to destroy it ourselves, otherwise it'll be retained in memory.
-      if (this._renderer.destroyNode) {
-        this._renderer.destroyNode(badgeElement);
-      }
-    }
-  }
-
   /**
-   * Gets the element into which the badge's content is being rendered.
-   * Undefined if the element hasn't been created (e.g. if the badge doesn't have content).
+   * Gets the element into which the badge's content is being rendered. Undefined if the element
+   * hasn't been created (e.g. if the badge doesn't have content).
    */
   getBadgeElement(): HTMLElement | undefined {
     return this._badgeElement;
   }
 
-  /** Injects a span element into the DOM with the content. */
-  private _updateTextContent(): HTMLSpanElement {
-    if (!this._badgeElement) {
+  ngOnInit() {
+    // We may have server-side rendered badge that we need to clear.
+    // We need to do this in ngOnInit because the full content of the component
+    // on which the badge is attached won't necessarily be in the DOM until this point.
+    this._clearExistingBadges();
+
+    if (this.content && !this._badgeElement) {
       this._badgeElement = this._createBadgeElement();
```

Comment:
IMHO: it's confusing that `_updateContent` sometimes handles creating the badge element, and sometimes the caller handles it. I spent a few minutes head scratching and thinking through the event lifecycle to figure this out. 

From the method's description: `  /** Update the text content of the badge.  */`. In this situation it would make more sense for `_updateContent` to do that one thing, and not do anything else. It's up to the caller to make sure the badge element is created.

I suggest making the caller handle creating the badge element, or making `_updateContent` create the badge element in all situations, or making a `_createBadgeElementIfNotExists` (final name depends on bike shedding) method to handle creating the badge element in a single place.

Reply:
I agree that the logic could be organized more cleanly here, but I want to avoid refactoring any more than I've already done here. As a middle ground, I've renamed `_updateContent` to _updateRenderedContent`, which I think does a better job of capturing that it may create a DOM element if it doesn't already exist.